onnxruntime/core/framework/kernel_registry_manager.cc

```diff
@@ -75,7 +75,7 @@ void KernelRegistryManager::RegisterKernelRegistry(std::shared_ptr<KernelRegistr
 bool KernelRegistryManager::HasImplementationOf(const Node& node, const std::string& provider_type) const {
   auto kernel_registries = GetKernelRegistriesByProviderType(provider_type);
   for (auto* kernel_registry : kernel_registries) {
-    if (kernel_registry->TryFindKernel(node, provider_type) != nullptr) {
+    if (kernel_registry && kernel_registry->TryFindKernel(node, provider_type) != nullptr) {
```

Comment:
>kernel_registry [](start = 8, length = 15)

There are no nulls in kernel_registries so this is an unnecessary check. 

   * @return It returns all the possible results. The returned value may contain garbage that doesn't belong to
   *         this provider. Caller should do the filtering. The returned value won't have no nullptrs.
   */
  std::vector<const KernelRegistry*> GetKernelRegistriesByProviderType(const std::string& type) const {

Reply:
TRT kernel is built by Compile function now, and it's not registered any more, so for TRT, kernel_registry is nullptr.

Reply:
Not following. KernelRegistryManager (correctly) doesn't allow registering a nullptr. Given that, I don't believe GetKernelRegistriesByProviderType can return any nullptr values. Is that not the case?

void KernelRegistryManager::RegisterKernelRegistry(std::shared_ptr<KernelRegistry> kernel_registry) {
  if (nullptr == kernel_registry) {
    return;
  }

Reply:
The nullptr of tensorrt is set in RegisterKernels() during session Initialize(). That's why GetKernelRegistriesByProviderType could return a nullptr.

Reply:
Can we fix RegisterKernels instead to not allow inserting nulls? Or is there a use case where you actually need a nullptr in there? 

If we fix it there, no other code needs the check on nullptr as the kernel registry maintains the invariant that there are no nullptr entries. 

---
In reply to: [262300042](https://github.com/Microsoft/onnxruntime/pull/382#discussion_r262300042) [](ancestors = 262300042)

Reply:
I fixed it in GetKernelRegistry(), where the nullptr returns. Now the kernel_registry check can be removed.